### PR TITLE
fix: prevent initial validation when the field is initialized as invalid (CP: 14)

### PR DESCRIPTION
**Pull request by roastedcpu**
Backports the following PRs as part of an EoD task:

- https://github.com/vaadin/web-components/pull/4138

### CODE REVIEW

File: test/email-field-validation.html [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    <link rel="import" href="../vaadin-text-area.html">
+    <link rel="import" href="../vaadin-text-field.html">
+    <link rel="import" href="../../iron-form/iron-form.html">
+    <link rel="import" href="helpers.html">
+</head>
+
+<body>
+<script>
+  ['vaadin-email-field'].forEach(el => {
+    describe(`${el}: initial validation`, () => {
```
Comment (vursen) [inline test/email-field-validation.html:22]:
```suggestion
    describe(`initial validation`, () => {
```

File: test/email-field-validation.html [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    <script src="../../web-component-tester/browser.js"></script>
+    <script src="../../webcomponentsjs/webcomponents-lite.js"></script>
+    <link rel="import" href="../../test-fixture/test-fixture.html">
+    <link rel="import" href="../vaadin-email-field.html">
+    <link rel="import" href="../vaadin-number-field.html">
+    <link rel="import" href="../vaadin-password-field.html">
+    <link rel="import" href="../vaadin-text-area.html">
+    <link rel="import" href="../vaadin-text-field.html">
+    <link rel="import" href="../../iron-form/iron-form.html">
```
Comment (vursen) [inline test/email-field-validation.html:15]: Please, remove the imports that aren't relevant to this test suite.

File: test/email-field-validation.html [hunk 3]
```diff
@@ -0,0 +1,64 @@
+<!doctype html>
+
+<head>
+    <meta charset="UTF-8">
+    <title>vaadin-text-field tests</title>
```
Comment (vursen) [inline test/email-field-validation.html:5]:
```suggestion
    <title>text-field validation tests</title>
```

File: test/text-field-validation.html [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    <link rel="import" href="../vaadin-password-field.html">
+    <link rel="import" href="../vaadin-text-area.html">
+    <link rel="import" href="../vaadin-text-field.html">
+    <link rel="import" href="../../iron-form/iron-form.html">
+    <link rel="import" href="helpers.html">
+</head>
+
+<body>
+<script>
+      var el = 'vaadin-text-field';
```
Comment (vursen) [inline test/text-field-validation.html:21]: Please use the same indentation in all tests.

File: test/email-field-validation.html [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  <script src="../../web-component-tester/browser.js"></script>
+  <script src="../../webcomponentsjs/webcomponents-lite.js"></script>
+  <link rel="import" href="../vaadin-email-field.html">
+  <link rel="import" href="helpers.html">
+</head>
+
+<body>
+<script>
+  var el = 'vaadin-email-field';
```
Comment (vursen) [inline test/email-field-validation.html:15]: It seems that there isn't much benefit in extracting tag names into a variable.

File: test/password-field-validation.html [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    });
+
+    it('should not validate by default', async() => {
+      document.body.appendChild(field);
+      await nextRender();
+      expect(validateSpy.called).to.be.false;
+    });
+
+    it('should not validate when the field has an initial value', async() => {
+      field.value = 'foo@example.com';
```
Comment (vursen) [inline test/password-field-validation.html:35]: Let's use the value from the original tests.

File: test/validation.html [hunk 7]
```diff
@@ -13,7 +13,7 @@
   <link rel="import" href="../vaadin-text-area.html">
   <link rel="import" href="../vaadin-text-field.html">
   <link rel="import" href="../../iron-form/iron-form.html">
-
+  <link rel="import" href="helpers.html">
```
Comment (vursen) [inline test/validation.html:16]: Let's remove unrelated changes from the PR.

File: test/password-field-validation.html (hunk 6 above)
Comment (vursen) [reply test/password-field-validation.html]: Thanks for fixing it for `password-field`. The comment also applies to the rest components.

File: test/email-field-validation.html [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  });
+</script>
+</body>
+
+
+
+
+
+
+
```
Comment (vursen) [inline test/email-field-validation.html:56]: Let's remove empty lines.

File: test/email-field-validation.html [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    beforeEach(() => {
+      field = document.createElement('vaadin-email-field');
+      validateSpy = sinon.spy(field, 'validate');
+    });
+
+    afterEach(() => {
+      field.remove();
+    });
+
+    it('should not validate by default', async() => {
```
Comment (vursen) [inline test/email-field-validation.html:27]:
```suggestion
    it('should not validate by default', async () => {
```

Comment (roastedcpu) [reply test/email-field-validation.html:27]: The linter complains: error  Unexpected space before function parentheses  space-before-function-paren
I checked the other files and the standard seems to be function(), async()

Comment (vursen) [reply test/email-field-validation.html:27]: Ah, I didn't notice that. It is fine then 👍

File: src/vaadin-text-field-mixin.html [hunk 10]
```diff
@@ -480,7 +480,7 @@
      */
     _valueChanged(newVal, oldVal) {
       // setting initial value to empty string, skip validation
-      if (newVal === '' && oldVal === undefined) {
+      if (oldVal === undefined) {
```
Comment (vursen) [inline src/vaadin-text-field-mixin.html:483]: The visual tests are failing for a good reason. This change skips not only validation but also value property propagation to the input element. Only validation should be skipped.

**Review by vursen — APPROVED** [review]
One visual test still failed apparently due to a sub-pixel rendering difference. We can ignore it. 

```
✘ *** text-area-scrolled-lumo text-area-scrolled [chrome]
```

**Current:**
chrome~current

**Ref:**
chrome~ref
